feat: openapi typescript client (#2885)
Redocly/redocly-cli: docs/@v2/guides/use-generated-client.md

| ---------------- | -------------------------------------------------------------------------------------------------- | -------------------------------------------------------- |
| `sdk`            | The typed client (default).                                                                        | none                                                     |
| `zod`            | `<output>.zod.ts` — [Zod](https://zod.dev) schemas + [validation middleware](#runtime-validation). | `zod` `^3.23 \|\| ^4`                                    |
| `tanstack-query` | `<output>.tanstack.ts` — [TanStack Query](https://tanstack.com/query) v5 factories.                | `@tanstack/<framework>-query` `^5`                       |
| `swr`            | `<output>.swr.ts` — [SWR](https://swr.vercel.app) hooks.                                           | `swr` `^2`                                               |
| `mock`           | `<output>.mocks.ts` — [MSW](https://mswjs.io) v2 handlers + `create<Schema>` factories.            | `msw` `^2` (+ `@faker-js/faker` for `--mock-data faker`) |

> | `transformers`   | `<output>.transformers.ts` — `transform<Name>` functions that parse wire dates to `Date`.          | none                                                     |

Reviewer: I am a bit confused by it. Do we really need it?

Reviewer: Should we also have a `docs` generator? which will generate a markdown documentation describing the client?

But also docs may be different for each generator so maybe it should be a separate `--with-docs` config?

Register with `use()` (sugar for `client.use()`); it accepts several at once:

```ts
import { use } from './client.ts';

use({
  onRequest: (ctx) => {
    // ctx.operation is { id, path, tags } — target by identity, not URL matching
    if (ctx.operation.tags.includes('Orders')) {
      ctx.headers['X-Idempotency-Key'] = crypto.randomUUID();
    }
  },
});
```

`onRequest` runs in registration order; `onResponse` runs in reverse (onion).
`onRequest` may mutate `ctx` (`url`, `method`, `headers`, and `body` — body edits are serialized and sent); `onResponse` may return a replacement `Response`.
`onError` (throw mode only) is threaded through each middleware.
`ctx.operation`'s fields are typed as the spec's literal unions (`OperationId`/`OperationPath`/`OperationTag`), so `ctx.operation.id === '…'` and `ctx.operation.tags.includes('…')` autocomplete, and a misspelled operation id fails compilation instead of silently never matching.
A header for a single call instead goes in that operation's trailing `init` argument.
Per-request headers merge lowest to highest: injected auth credentials, then typed header parameters, then the caller's `init.headers` — the caller always wins.

> `use()` appends to the middleware chain, composing with any already-registered or baked-in middleware.

Reviewer: I don't like the "baked-in" terms everywhere. Should we change it to "pre-configured"? or something like that?

`configure({ middleware: [...] })` replaces the whole chain — use it to reset, but prefer `use()` to add to existing (including [publisher-baked](#publisher-defaults)) middleware.

See the [`configure-and-middleware` example](https://github.com/Redocly/redocly-cli/tree/main/tests/e2e/generate-client/examples/configure-and-middleware) for a runnable version.

## Publisher defaults

> The middleware above is composed by the **consumer**.

Reviewer: I think this piece of docs does not belong here. It is not about usage the generated client

> If you **publish an SDK**, bake defaults in at generation time with `--setup <file>`.

Reviewer: ```suggestion
If you **publish an SDK** you can pre-configure default middlewares at generation time with `--setup <file>`.
```

Reviewer: @Marshevskyy see here

The setup module imports its contract from `@redocly/client-generator` (so it resolves and is unit-testable) and default-exports `defineClientSetup({ config, middleware })`:

```ts
// client-setup.ts

> import { defineClientSetup, type RequestContext } from '@redocly/client-generator';

Reviewer: Will it work with "inline" runtime?

Reviewer: I would not require any imports and just expect this file to export a js object. do we really need the function call? We can wire the function call ourself

Author: yes it will work inline runtime. function call is not required

Reviewer: so should we update the docs if the function call is not required?

export default defineClientSetup({
  config: { serverUrl: 'https://api.acme.com', retry: { retries: 3 } },
  middleware: [
    {
      onRequest: (ctx: RequestContext) => {
        ctx.headers['X-Acme-SDK'] = '1.4.0';
      },
    },
  ],
});
```

```sh
redocly generate-client openapi.yaml --output src/api/client.ts --setup ./client-setup.ts
```

> The baked block runs before the consumer's own setup.

Reviewer: ```suggestion
The pre-configured block runs before the consumer's own setup.
```

Author: I would keep baked since this specifically conveys "embedded at generation time, travels inside the file.", do you still want to change name?

Resume by passing the advance param in the initial args — iteration starts from there instead of the beginning.
Abort by passing an `AbortSignal`, forwarded to every page request:

```ts
const controller = new AbortController();
for await (const page of client.listOrders.pages(
  { params: { cursor: 'c2' } }, // start from a saved cursor (or offset/page number)
  { signal: controller.signal }
)) {
  // …
}
```

A failed page always aborts iteration by throwing `ApiError`, even on an `--error-mode result` client — there `.pages()` yields raw pages (not `{ data, error, response }` envelopes; only the one-shot call keeps the envelope) and the throw-mode-only `onError` middleware hook is not invoked.

For shapes the built-in styles don't cover — for example a cursor that travels in the request body or a header — page with a small hand-written helper over the generated call, which stays fully typed end to end (see the [`custom-pagination` example](https://github.com/Redocly/redocly-cli/tree/main/tests/e2e/generate-client/examples/custom-pagination)).

> ## Custom generators

Reviewer: This one also doen'st belong here
